wrapper/src/main/java/software/amazon/jdbc/util/SlidingExpirationCache.java

```diff
@@ -0,0 +1,261 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.util;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+public class SlidingExpirationCache<K, V> {
+  private final Map<K, CacheItem> cache = new ConcurrentHashMap<>();
+  private long cleanupIntervalNanos = TimeUnit.MINUTES.toNanos(10);
+  private final AtomicLong cleanupTimeNanos = new AtomicLong(System.nanoTime() + cleanupIntervalNanos);
+  private final ShouldDisposeFunc<V> shouldDisposeFunc;
+  private final ItemDisposalFunc<V> itemDisposalFunc;
+
+  /**
+   * A cache that periodically cleans up expired entries. Fetching an expired entry marks that entry
+   * as not-expired and renews its expiration time.
+   */
+  public SlidingExpirationCache() {
+    this.shouldDisposeFunc = null;
+    this.itemDisposalFunc = null;
+  }
+
+  /**
+   * A cache that periodically cleans up expired entries. Fetching an expired entry marks that entry
+   * as not-expired and renews its expiration time.
+   *
+   * @param shouldDisposeFunc a function defining the conditions under which an expired entry should
+   *                          be cleaned up when we hit the cleanup time
+   * @param itemDisposalFunc  a function that will be called on any item that meets the cleanup
+   *                          criteria at cleanup time. The criteria for cleanup is that the item
+   *                          is both expired and marked for cleanup via a call to
+   *                          shouldDisposeFunc.
+   */
+  public SlidingExpirationCache(
+      final ShouldDisposeFunc<V> shouldDisposeFunc,
+      final ItemDisposalFunc<V> itemDisposalFunc) {
+    this.shouldDisposeFunc = shouldDisposeFunc;
+    this.itemDisposalFunc = itemDisposalFunc;
+  }
+
+  /**
+   * In addition to performing the logic defined by {@link Map#computeIfAbsent}, cleans up expired
+   * entries if we have hit cleanup time. If an expired entry is requested and we have not hit
+   * cleanup time or {@link ShouldDisposeFunc} indicated the entry should not be closed, the entry
+   * will be marked as non-expired.
+   *
+   * @param key                the key with which the specified value is to be associated
+   * @param mappingFunction    the function to compute a value
+   * @param itemExpirationNano the expiration time of the new or renewed entry
+   * @return the current (existing or computed) value associated with the specified key, or null if
+   *     the computed value is null
+   */
+  public V computeIfAbsent(
```

Comment:
Should we remove similar method (and sliding expiration related functionality) from CacheMap<> ?